packages/lib/src/interactions/SelectionTool.tsx

```diff
@@ -130,49 +123,63 @@ function SelectionTool(props: Props) {
     'Escape',
     () => {
       startEvtRef.current = undefined;
-      setSelectionState({ selection: undefined, isCancelled: true });
+      setRawSelection(undefined);
     },
     [],
     { event: 'keydown' }
   );
 
-  useEffect(() => {
-    const { selection, isCancelled } = selectionState;
+  // Compute effective selection
+  const selection = useMemo(
+    () => rawSelection && transformSelectionRef.current(rawSelection),
+    [rawSelection, transformSelectionRef]
+  );
+
+  // Keep track of previous effective selection state
+  const prevSelection = usePrevious(selection);
 
+  useEffect(() => {
     if (selection) {
+      assertDefined(rawSelection);
+
       // Previous selection was undefined and current selection is now defined => selection has started
-      if (!prevSelectionState?.selection) {
-        onSelectionStartRef.current?.(selection);
+      if (!prevSelection) {
+        onSelectionStartRef.current?.();
       }
 
-      // Either way, current selection is defined, so invoke change callback
+      // Either way, current selection is defined, so invoke change callback with effective selection object
       onSelectionChangeRef.current?.(
-        isModifierKeyPressed ? selection : undefined // don't pass selection object if user is not pressing modifier key
+        isModifierKeyPressed ? selection : undefined, // don't pass selection object if user is not pressing modifier key
+        rawSelection
```

Comment:
Note that contrary to the effective selection, the raw selection is always passed to the callback.